os/arch/arm/src/armv7-a/arm_cpuhead.S

```diff
@@ -51,6 +51,8 @@
 #include "chip.h"
 #include "up_internal.h"
 
```

Comment:
Is this needed?

Reply:
Yes its needed as STACK_COLOR is used in this file.

Reply:
So, you moved the definition from up_internal.h to arch.h, so that it is callable from kernel/sched as well?

Reply:
yes

Reply:
arm_cpuhead.S is for secondary cores booting sequence (ie. core index 2, 3, 4). Does it mean this stack overflow detection only support multicore scenario?
How about single core case, the primary core enters through arm_head.S

Reply:
Have you conducted some test scenario on this feature (ie. purposely causing stack overflow), on both single core and multicore cases?
I think more information should be added when stack overflow occurs.
1) task/thread pid and name
2) allocated stack size and overflowed value
could you add above and share the test result here?

Reply:
> arm_cpuhead.S is for secondary cores booting sequence (ie. core index 2, 3, 4). Does it mean this stack overflow detection only support multicore scenario? How about single core case, the primary core enters through arm_head.S

Stack coloration for idle task of secondary cores is done in this file. 
Stack coloration for idle task of core 0 is done in os_start -> up_initialstate -> up_stack_color.
So, stack overflow detection will work for all cases.

Reply:
> Have you conducted some test scenario on this feature (ie. purposely causing stack overflow), on both single core and multicore cases? I think more information should be added when stack overflow occurs.
> 
> 1. task/thread pid and name
> 2. allocated stack size and overflowed value
> 
> Could you add above and share the test result here?

The stack overflow detection works as below:
- During context switch we check if stack has overflowed
- If overflow is detected, we immediately call PANIC
- The assert handler displays complete details about the task and its stack size, consumed size, etc.. 
- The contents of the stack is also dumped. 

However, there are limitations to this detection method. The stack detection is possible only for sequential stack updates. However, if a thread allocates a large array beyond the stack boundary and starts to write into the array from the lower address, then stack overflow is not detected immediately. The overflow is only detected when the stack boundary is breached and the COLOR value is overwritten.

Reply:
Thanks for explanation, it is clear to me~